Add switcher s12 support

YogevBokobza opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Switcher Runner S12 support

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/35014

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -72,7 +83,15 @@ def __init__(
         self._cover_id = cover_id
 
-        self._attr_unique_id = f"{coordinator.device_id}-{coordinator.mac_address}"
+        if (
```
It would be better to have a base class with the common functionality and two subclasses, similar to:
https://github.com/home-assistant/core/blob/e54031e318ff7af281ae399d6de9743411bf31a9/homeassistant/components/switcher_kis/switch.py#L80

So you would end up with
- `SwitcherBaseCoverEntity`
- `SwitcherSingleCoverEntity`
- `SwitcherDualCoverEntity`

The subclass will only implement changed parts

YogevBokobza replied:
It is done.. I tried to separate __init__ as well but it did make issues..

thecode replied:
Please submit the separate init and we will fix the issues in the PR review.

YogevBokobza replied:
It is done. You can see current issue.

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -133,3 +152,23 @@ async def async_set_cover_position(self, **kwargs: Any) -> None:
+
+
+class SwitcherSingleCoverEntity(SwitcherBaseCoverEntity):
+    """Representation of a Switcher single cover entity."""
```
Update method should be moved from the base class to here as it is not common for both classes:
https://github.com/home-assistant/core/blob/b84ce69ab4254f28c0a0f146eced8bb0e5198004/homeassistant/components/switcher_kis/cover.py#L104-L110

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -63,33 +74,12 @@ class SwitcherCoverEntity(SwitcherEntity, CoverEntity):
         | CoverEntityFeature.STOP
     )
 
```
```suggestion
    _cover_id: int
```
If you add it to the base class it will resolve the warning

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -63,33 +74,12 @@ class SwitcherCoverEntity(SwitcherEntity, CoverEntity):
-
-        self._update_data()
-
     @callback
```
Add empty method:
```python
    def _update_data(self) -> None:
        """Update data from device."""
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/light.py:
```diff
@@ -87,8 +95,8 @@ def is_on(self) -> bool:
 
-        data = cast(SwitcherSingleShutterDualLight, self.coordinator.data)
-        return bool(data.lights[self._light_id] == DeviceState.ON)
+        data = cast(SwitcherDualShutterSingleLight, self.coordinator.data)
+        return bool(data.lights == DeviceState.ON)
```
Please move this to `SwitcherSingleLightEntity` class

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/light.py:
```diff
@@ -127,3 +135,20 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        if self.control_result is not None:
+            return self.control_result
+
```
```suggestion
        """Return True if entity is on."""
          super().is_on()
```

YogevBokobza replied:
It is done.

thecode replied:
This was just an example not something you should commit (should be something like `return `super().is_on() or ...`, however looking at the final result I think it would be better to remove this from the base class and add it in the subclasses.

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/light.py:
```diff
@@ -127,3 +135,20 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+class SwitcherSingleLightEntity(SwitcherBaseLightEntity):
+    """Representation of a Switcher single light entity."""
+
+
```
```python
    @property
    def is_on(self) -> bool:
        """Return True if entity is on."""
        super().is_on
        data = cast(SwitcherDualShutterSingleLight, self.coordinator.data)
        return bool(data.lights == DeviceState.ON)
```

YogevBokobza commented:
@thecode 
The last changes requested caused issues with the tests and lining. How to proceed?

thecode commented:
@YogevBokobza please go over https://developers.home-assistant.io/docs/core/entity/#entity-naming and make sure entity names are correct. Before this PR there where only single cover device or dual light devices, you now add single and dual to both classes.
Entity names should not include an ID for a single light/cover device and include if there are multiple.

YogevBokobza commented:
> @YogevBokobza please go over https://developers.home-assistant.io/docs/core/entity/#entity-naming and make sure entity names are correct. Before this PR there where only single cover device or dual light devices, you now add single and dual to both classes. Entity names should not include an ID for a single light/cover device and include if there are multiple.

I am not sure I got your point.. Can you show me which lines of code you are referring to?

YogevBokobza committed:
switcher fix for dual and single

YogevBokobza commented:
> @YogevBokobza please go over https://developers.home-assistant.io/docs/core/entity/#entity-naming and make sure entity names are correct. Before this PR there where only single cover device or dual light devices, you now add single and dual to both classes. Entity names should not include an ID for a single light/cover device and include if there are multiple.

I think I understand now.. Please see recent code if this is what you mean..

YogevBokobza commented:
This is what S11 looks like:
![s11](https://github.com/user-attachments/assets/d7cadd20-b2fd-459a-85ae-8b6051cec8f3)
IDS:
cover.switcher_runner_6cf5
light.switcher_runner_6cf5_light_1
light.switcher_runner_6cf5_light_2

This is what S12 looks like:
![S12](https://github.com/user-attachments/assets/752faf90-f8fb-45b8-b324-37274fc42739)
light.switcher_run_plus_a9be_light
cover.switcher_run_plus_a9be
cover.switcher_run_plus_a9be_2

@thecode Looks ok?

YogevBokobza committed:
Update light.py

YogevBokobza commented:
Rebased to get this [change](https://github.com/home-assistant/core/pull/128750)

YogevBokobza commented:
@thecode Should we go and merge this PR and afterwards bump aioswitcher to align position, direction and light to array or we pause this PR and make that bump and afterwards rebase this PR?

thecode commented:
> @thecode Should we go and merge this PR and afterwards bump aioswitcher to align position, direction and light to array or we pause this PR and make that bump and afterwards rebase this PR?

Lets first bump aioswitcher and rebase this PR

thecode commented:
> This is what S11 looks like: ![s11](https://private-user-images.githubusercontent.com/22839127/373956912-d7cadd20-b2fd-459a-85ae-8b6051cec8f3.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5NDc3NTcsIm5iZiI6MTcyOTk0NzQ1NywicGF0aCI6Ii8yMjgzOTEyNy8zNzM5NTY5MTItZDdjYWRkMjAtYjJmZC00NTlhLTg1YWUtOGI2MDUxY2VjOGYzLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDEyNTczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1MzgyNzY3OTFjZDhjOWEwYjFiZmRhYTc4Nzk1OWMzZDg0NDk3M2ZhYmU5NzJjYmZjZDcwN2I3NzBmYzE5MzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jKt8pMNpMjfeGmnSveiLNglAIqAnbqQMeuImNACf8EM) 
IDS: 
cover.switcher_runner_6cf5 
light.switcher_runner_6cf5_light_1 
light.switcher_runner_6cf5_light_2
> 
> This is what S12 looks like: ![S12](https://private-user-images.githubusercontent.com/22839127/373956954-752faf90-f8fb-45b8-b324-37274fc42739.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5NDc3NTcsIm5iZiI6MTcyOTk0NzQ1NywicGF0aCI6Ii8yMjgzOTEyNy8zNzM5NTY5NTQtNzUyZmFmOTAtZjhmYi00NWI4LWIzMjQtMzcyNzRmYzQyNzM5LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDEyNTczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxODZlOGYwYmE3NmQwZjI5OTY5YTE1ZGMwNWRlNTkwZGU2Njg0NGUwMjA1MWQwNjIwZDdhNDlkYTlkNDg2NTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pzcyi9Gk88mJlhyB2b4LpF3SNa6q26tt_XJ3V165cbM) 
light.switcher_run_plus_a9be_light 
cover.switcher_run_plus_a9be 
cover.switcher_run_plus_a9be_2
> 
> @thecode Looks ok?

No, first something is wrong with the entities IDs although I didn't directly see in the implementation why, maybe you should remove the device and add it again to generate new entity IDs. The cover on the 2nd example should be:
cover.switcher_run_plus_a9be 
cover.switcher_run_plus_a9be_cover_1
cover.switcher_run_plus_a9be_cover_2

I also suggest to change the name from "Runner" to "Shutter" or "Cover" as "Runner" is the device name by the manufacture and a "Runner" device may have both shutter control and light control.

Please fix and rebase so we can finish the review.

YogevBokobza commented:
> > This is what S11 looks like: ![s11](https://private-user-images.githubusercontent.com/22839127/373956912-d7cadd20-b2fd-459a-85ae-8b6051cec8f3.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5NDc3NTcsIm5iZiI6MTcyOTk0NzQ1NywicGF0aCI6Ii8yMjgzOTEyNy8zNzM5NTY5MTItZDdjYWRkMjAtYjJmZC00NTlhLTg1YWUtOGI2MDUxY2VjOGYzLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDEyNTczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1MzgyNzY3OTFjZDhjOWEwYjFiZmRhYTc4Nzk1OWMzZDg0NDk3M2ZhYmU5NzJjYmZjZDcwN2I3NzBmYzE5MzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jKt8pMNpMjfeGmnSveiLNglAIqAnbqQMeuImNACf8EM)
> > IDS:
> > cover.switcher_runner_6cf5
> > light.switcher_runner_6cf5_light_1
> > light.switcher_runner_6cf5_light_2
> > This is what S12 looks like: ![S12](https://private-user-images.githubusercontent.com/22839127/373956954-752faf90-f8fb-45b8-b324-37274fc42739.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5NDc3NTcsIm5iZiI6MTcyOTk0NzQ1NywicGF0aCI6Ii8yMjgzOTEyNy8zNzM5NTY5NTQtNzUyZmFmOTAtZjhmYi00NWI4LWIzMjQtMzcyNzRmYzQyNzM5LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI2VDEyNTczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxODZlOGYwYmE3NmQwZjI5OTY5YTE1ZGMwNWRlNTkwZGU2Njg0NGUwMjA1MWQwNjIwZDdhNDlkYTlkNDg2NTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pzcyi9Gk88mJlhyB2b4LpF3SNa6q26tt_XJ3V165cbM)
> > light.switcher_run_plus_a9be_light
> > cover.switcher_run_plus_a9be
> > cover.switcher_run_plus_a9be_2
> > @thecode Looks ok?
> 
> No, first something is wrong with the entities IDs although I didn't directly see in the implementation why, maybe you should remove the device and add it again to generate new entity IDs. The cover on the 2nd example should be: `light.switcher_run_plus_a9be_light` `cover.switcher_run_plus_a9be_runner_1` `cover.switcher_run_plus_a9be_runner_2`
> 
> I also suggest to change the name from "Runner" to "Shutter" or "Cover" as "Runner" is the device name by the manufacture and a "Runner" device may have both shutter control and light control.
> 
> Please fix and rebase so we can finish the review.

Sure will do.
Rebase is harder because all the conflicts. I am trying to solve it..

YogevBokobza committed:
switcher add s12 support
# Conflicts:
#	homeassistant/components/switcher_kis/cover.py
#	homeassistant/components/switcher_kis/light.py
#	homeassistant/components/switcher_kis/manifest.json
#	requirements_all.txt
#	requirements_test_all.txt
#	tests/components/switcher_kis/consts.py
#	tests/components/switcher_kis/test_cover.py
#	tests/components/switcher_kis/test_light.py

YogevBokobza committed:
bump aioswitcher to 4.1.0
# Conflicts:
#	homeassistant/components/switcher_kis/manifest.json
#	requirements_all.txt
#	requirements_test_all.txt

YogevBokobza committed:
fix based on requested changes
# Conflicts:
#	homeassistant/components/switcher_kis/cover.py

YogevBokobza committed:
fix linting
# Conflicts:
#	homeassistant/components/switcher_kis/light.py

YogevBokobza committed:
fix linting issue
# Conflicts:
#	homeassistant/components/switcher_kis/cover.py
#	homeassistant/components/switcher_kis/light.py

YogevBokobza committed:
fix based on requested changes
# Conflicts:
#	homeassistant/components/switcher_kis/cover.py

YogevBokobza committed:
fix based on requested changes
# Conflicts:
#	homeassistant/components/switcher_kis/light.py

YogevBokobza committed:
fix light.py
# Conflicts:
#	homeassistant/components/switcher_kis/light.py

YogevBokobza committed:
fix after rebase

YogevBokobza committed:
Update cover.py, strings.json, and test_cover.py

YogevBokobza commented:
@thecode 
So right now after rebase and fix we have those cases:
Normal Runner(Normal, Mini etc) - will look like this:
![image](https://github.com/user-attachments/assets/f245fe28-0fe4-488a-9825-4c12aac9b9cb)

IDs:
cover.balcony_runner
* The name here comes from the Switcher App setup.
** Leave it as it was before

Runner S11 - will look like this:
![image](https://github.com/user-attachments/assets/1edf2da5-4393-472b-995b-f96868785d5f)

IDs:
cover.switcher_runner_6cf5
light.switcher_runner_6cf5_light_1
light.switcher_runner_6cf5_light_2

Runner S12 - will look like this:
![image](https://github.com/user-attachments/assets/58f02527-59d1-402b-883a-4b7ccdd696ba)

IDs:
cover.switcher_run_plus_a9be_cover_1
cover.switcher_run_plus_a9be_cover_2
light.switcher_run_plus_a9be_light

I think that:
1. We should rename "DualCover" and "DualLight" to "ComplexCover" and "ComplexLight" so this will support the later 3 lights modules
2. I didn't like the S11 cover to get the actual device name (which is not what it is on the Switcher App), instead, I think that S11 should move to the "ComplexCover" and name it Cover


What do you think?

YogevBokobza commented:
> @thecode So right now after rebase and fix we have those cases: Normal Runner(Normal, Mini etc) - will look like this: ![image](https://private-user-images.githubusercontent.com/22839127/380488716-f245fe28-0fe4-488a-9825-4c12aac9b9cb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwMzAyNDEsIm5iZiI6MTczMDAyOTk0MSwicGF0aCI6Ii8yMjgzOTEyNy8zODA0ODg3MTYtZjI0NWZlMjgtMGZlNC00ODhhLTk4MjUtNGMxMmFhYzliOWNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI3VDExNTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0YzI2ZGNlOTViY2RmMmExYmNkOGRmZjM2NjBmYjE5ZmEyZDgwZjFhN2Q1NDY0NjQ0ZTZlZTU3OWY4MzdkYTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vnI44t337DhFjOOEbKmGW8oKy5XafkXHRSpPphLMk_Y)
> 
> IDs: cover.balcony_runner NOTE1: The name here comes from the Switcher App setup. NOTE2: Leave it as it was before
> 
> Runner S11 - will look like this: ![image](https://private-user-images.githubusercontent.com/22839127/380488780-1edf2da5-4393-472b-995b-f96868785d5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwMzAyNDEsIm5iZiI6MTczMDAyOTk0MSwicGF0aCI6Ii8yMjgzOTEyNy8zODA0ODg3ODAtMWVkZjJkYTUtNDM5My00NzJiLTk5NWItZjk2ODY4Nzg1ZDVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI3VDExNTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiYTNlYWU1YWY1NDY0YzQ4OTU3NTc3NDkxYTk2ZDEyYWFlYTcyMjM0MmRiMjlkZjJjNjM4MjY1Y2JjMjI3MmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qdyIBqkZwZEXmqvZG7r-awh0czbQ5iqO8HBURWPYkA0)
> 
> IDs: cover.switcher_runner_6cf5 light.switcher_runner_6cf5_light_1 light.switcher_runner_6cf5_light_2
> 
> Runner S12 - will look like this: ![image](https://private-user-images.githubusercontent.com/22839127/380488857-58f02527-59d1-402b-883a-4b7ccdd696ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwMzAyNDEsIm5iZiI6MTczMDAyOTk0MSwicGF0aCI6Ii8yMjgzOTEyNy8zODA0ODg4NTctNThmMDI1MjctNTlkMS00MDJiLTg4M2EtNGI3Y2NkZDY5NmJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI3VDExNTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3ODhjNDNiNGE3OTgxZTFhZTBlOGJhYjk0NDBhNGIxNWI4YWMwMzU0ZjQyYzVjMWQ5YTcxZTIwZmZlOGU2OGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cDcT3BoygPfibGtyZx8u9HLYoWTguMQStrSFP1XVe1U)
> 
> IDs: cover.switcher_run_plus_a9be_cover_1 cover.switcher_run_plus_a9be_cover_2 light.switcher_run_plus_a9be_light
> 
> I think that:
> 
> 1. We should rename "DualCover" and "DualLight" to "ComplexCover" and "ComplexLight" so this will support the later 3 lights modules
> 2. I didn't like the S11 cover to get the actual device name (which is not what it is on the Switcher App), instead, I think that S11 should move to the "ComplexCover" and name it Cover
> 
> What do you think?

Regarding that, After some testing, I think that if we want to accomplish (1) and (2) we need to change to:
`class SwitcherDualCoverEntity` will be split into 2:  `class SwitcherComplexSingleCoverEntity` and `class SwitcherComplexMultiCoverEntity` (Same will go for light)

So creating the entities will look like this:
```
        if coordinator.data.device_type.category == DeviceCategory.SHUTTER:
            async_add_entities([SwitcherSingleCoverEntity(coordinator, 0)])
        if (
            coordinator.data.device_type.category
            == DeviceCategory.SINGLE_SHUTTER_DUAL_LIGHT
        ):
            async_add_entities([SwitcherComplexSingleCoverEntity(coordinator, 0)])
        if (
            coordinator.data.device_type.category
            == DeviceCategory.DUAL_SHUTTER_SINGLE_LIGHT
        ):
            async_add_entities([SwitcherComplexMultiCoverEntity(coordinator, 0)])
            async_add_entities([SwitcherComplexMultiCoverEntity(coordinator, 1)])
```
Also adding `single_cover` to strings..

@thecode  What do you think?

thecode commented:
> @thecode What do you think?

Lets try to finish this PR and not start from scratch. Please address the requested changes. We can rename classes or add changes later or in future PRs.

YogevBokobza commented:
> > @thecode What do you think?
> 
> Lets try to finish this PR and not start from scratch. Please address the requested changes. We can rename classes or add changes later or in future PRs.

I already did address the requested changes here: https://github.com/home-assistant/core/pull/127277#issuecomment-2439968294

thecode commented on homeassistant/components/switcher_kis/light.py:
```diff
@@ -123,3 +115,44 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+class SwitcherDualLightEntity(SwitcherBaseLightEntity):
+    """Representation of a Switcher dual light entity."""
+
+    _attr_translation_key = "light"
```
Please look at the logic used in the cover, you don't need to translate the light for a single entity.

YogevBokobza replied:
It is done, but now the light will get the device name which is not what is shown in the Switcher App. (Just like the cover in the S11 device)

Now it will be like this:
Runner S11:
cover.switcher_runner_6cf5
light.switcher_runner_6cf5_light_1
light.switcher_runner_6cf5_light_2

Runner S12:
cover.switcher_run_plus_a9be_cover_1
cover.switcher_run_plus_a9be_cover_2
light.switcher_run_plus_a9be

thecode commented on homeassistant/components/switcher_kis/strings.json:
```diff
@@ -43,7 +43,15 @@
+      }
+    },
     "light": {
+      "single_light": {
+        "name": "Light"
+      },
```
```suggestion
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/light.py:
```diff
@@ -123,3 +115,44 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        light_id: int = 0,
```
```suggestion
        light_id: int,
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/light.py:
```diff
@@ -123,3 +115,44 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        light_id: int = 0,
```
```suggestion
        light_id: int,
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -137,3 +130,44 @@ async def async_set_cover_position(self, **kwargs: Any) -> None:
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        cover_id: int = 0,
```
```suggestion
        cover_id: int,
```

YogevBokobza replied:
It is done.

thecode commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -137,3 +130,44 @@ async def async_set_cover_position(self, **kwargs: Any) -> None:
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        cover_id: int = 0,
```
```suggestion
        cover_id: int,
```

YogevBokobza replied:
It is done.

YogevBokobza committed:
fix based on requested changes

joostlek commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -44,37 +44,30 @@ def async_add_cover(coordinator: SwitcherDataUpdateCoordinator) -> None:
             DeviceCategory.SINGLE_SHUTTER_DUAL_LIGHT,
         ):
-            async_add_entities([SwitcherCoverEntity(coordinator, 0)])
+            async_add_entities([SwitcherSingleCoverEntity(coordinator, 0)])
+        if (
+            coordinator.data.device_type.category
+            == DeviceCategory.DUAL_SHUTTER_SINGLE_LIGHT
+        ):
+            async_add_entities([SwitcherDualCoverEntity(coordinator, 0)])
+            async_add_entities([SwitcherDualCoverEntity(coordinator, 1)])
```
We should not call `async_add_entities` more than once at a time. Please collect the entities in a list and add them in one go

joostlek commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -137,3 +130,44 @@ async def async_set_cover_position(self, **kwargs: Any) -> None:
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        cover_id,
```
Can we type this

joostlek commented on homeassistant/components/switcher_kis/cover.py:
```diff
@@ -137,3 +130,44 @@ async def async_set_cover_position(self, **kwargs: Any) -> None:
+    def __init__(
+        self,
+        coordinator: SwitcherDataUpdateCoordinator,
+        cover_id,
```
idem

joostlek reviewed (changes requested):
Same goes for the light.py

joostlek committed:
Fix latest things

joostlek committed:
Fix latest things

thecode committed:
Merge branch 'dev' into switcher-add-s12-support

thecode reviewed (approved):
Thanks @YogevBokobza and @joostlek 👍
